### PR TITLE
Add instrumentation tests to CI

### PR DESCRIPTION
**Pull request by niltonvasques**
- Refactor file downloader mocks in instrumented tests
- Fix gifLoadingTest
- Add an android emulator as base image
- Run instrumentation tests

[To run the X86 architecture, you will need to have kvm installed on the host machine, and you will have to run the image with the --privileged option. Running the emulator for Android 16 (X86)](https://github.com/softsam/docker-android-emulator)

[How to Setup drone build to run privileged builds](http://readme.drone.io/usage/build_test/#privileged:fb92aa3346185c57f15afda861d465a3)
### Waiting emulator boot finish

``` bash
#!/bin/bash

adb wait-for-device

A=$(adb shell getprop sys.boot_completed | tr -d '\r')

while [ "$A" != "1" ]; do
        sleep 2
        A=$(adb shell getprop sys.boot_completed | tr -d '\r')
done
```
### Checking if the device is locked

``` ruby
# Switches on the android devices screen if it isn’t already on.
if `adb shell dumpsys input_method | grep mInteractive`.include? 'false'
  `adb shell input keyevent KEYCODE_POWER`
end

# Pre lollipop devices respond to this command a bit differently.
if `adb shell dumpsys input_method | grep mScreenOn`.include? 'false'
  `adb shell input keyevent KEYCODE_POWER`
end
```
### Running drone build locally
- [Install](http://readme.drone.io/devs/cli/#install:b659b046131d4024ab5e2d3675716bf0) drone CLI
  
  ```
  curl http://downloads.drone.io/drone-cli/drone_linux_amd64.tar.gz | tar zx
  sudo install -t /usr/local/bin drone
  ```
- Run drone build inside project dir
  
  ```
  drone exec --trusted
  ```

### CODE REVIEW

**Review by fpbitencourt — APPROVED** [review]
:-1:

**Comment by niltonvasques** [comment]
@raulccabreu Do you can enable trusted settings in drone configs? We need this to run privileged containers in the builds and support KVM. Take a look in the PR infos.

**Comment by raulccabreu** [comment]
The trusted options is now enabled.

**Comment by raulccabreu** [comment]
Only to remeber, the first comment will go to docker image readme??

File: .drone.yml [hunk 1]
```diff
@@ -1,4 +1,14 @@
 build:
-  image: catbagdev/android-env
+  image: niltonvasques/android-emulator:1.3
```
Comment (raulccabreu) [inline .drone.yml:2]: To remeber...change to central image repo...

**Comment by niltonvasques** [comment]
Yes! Part of that comment will go into docker image readme and other part will go in readme from that project.

File: app/src/androidTest/java/br/com/catbag/gifreduxsample/ui/giflist/GifListActivityTest.java [hunk 2]
```diff
@@ -94,35 +99,16 @@ public void pauseGifTest() {
         onView(withId(R.id.gif_image))
                 .perform(click());
 
+        View image = mActivityTestRule.getActivity().findViewById(R.id.gif_image);
+        Log.i("pauseGifTest", "before "+image.getWidth()+" W"+ image.getHeight()+ " H");
+        sleep(2000);
```
Comment (raulccabreu) [inline app/src/androidTest/java/br/com/catbag/gifreduxsample/ui/giflist/GifListActivityTest.java]: To remeber, change all sleeps to idling resource....probably a issue to be opened

```
// Make sure Espresso does not time out
IdlingPolicies.setMasterPolicyTimeout(waitingTime * 2, TimeUnit.MILLISECONDS);
IdlingPolicies.setIdlingResourceTimeout(waitingTime * 2, TimeUnit.MILLISECONDS);

// Now we wait
IdlingResource idlingResource = new ElapsedTimeIdlingResource(waitingTime);
Espresso.registerIdlingResources(idlingResource);

// Stop and verify
onView(withId(R.id.toggle_button))
    .check(matches(withText(R.string.stop)))
    .perform(click());
onView(withId(R.id.result))
    .check(matches(withText(success ? R.string.success: R.string.failure)));

// Clean up
Espresso.unregisterIdlingResources(idlingResource);
```

**Review by raulccabreu — CHANGES_REQUESTED** [review]

File: app/src/main/java/br/com/catbag/gifreduxsample/ui/GifListActivity.java [hunk 3]
```diff
@@ -27,7 +27,7 @@
     private GifActionCreator mActionCreator = GifActionCreator.getInstance();
 
     //Views
-    private GlideWrapper mGlideWrapper;
+    private GifWrapper mGlideWrapper;
```
Comment (raulccabreu) [inline app/src/main/java/br/com/catbag/gifreduxsample/ui/GifListActivity.java]: Glide reference...

File: app/src/main/java/br/com/catbag/gifreduxsample/ui/GifListActivity.java [hunk 4]
```diff
@@ -97,7 +97,7 @@ public void onStateChanged(AppState appState) {
         Anvil.render();
     }
 
-    public GlideWrapper getGlideWrapper() {
+    public GifWrapper getGlideWrapper() {
```
Comment (raulccabreu) [inline app/src/main/java/br/com/catbag/gifreduxsample/ui/GifListActivity.java]: Glide reference....

File: app/src/main/java/br/com/catbag/gifreduxsample/ui/wrappers/GifWrapper.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+public class GifWrapper {
+
+    private static final String TAG = "GifWrapper";
+    private String mLocalPath;
+
+    private GifImageView mImageView;
+    private GifDrawable mDrawable;
+
+    private GlideLoadListener mLoadListener;
```
Comment (raulccabreu) [inline app/src/main/java/br/com/catbag/gifreduxsample/ui/wrappers/GifWrapper.java]: Glide ref

File: app/src/main/java/br/com/catbag/gifreduxsample/ui/wrappers/GifWrapper.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            if (mExceptionListener != null) mExceptionListener.onException(e);
+        }
+    }
+
+    public GifWrapper onLoaded(GlideLoadListener listener) {
+        mLoadListener = listener;
+        return this;
+    }
+
+    public GifWrapper onException(GlideExceptionListener listener) {
```
Comment (raulccabreu) [inline app/src/main/java/br/com/catbag/gifreduxsample/ui/wrappers/GifWrapper.java]: Glide ref

**Review by raulccabreu — CHANGES_REQUESTED** [review]

File: .drone.yml (hunk 1 above)
Comment (niltonvasques) [reply .drone.yml:2]: We can change the image, when we need change something in the image.

**Review by raulccabreu — APPROVED** [review]

**Pull request merged**